docs/writing-policies/go/04-validation.md

```diff
@@ -8,331 +8,420 @@ title: ""
+			podLabels: map[string]string{
+				"cc-center": "team-123",
+			},
+			deniedLabels: []string{"hello"},
+			constrainedLabels: map[string]string{
+				"cc-center": `team-\d+`,
+			},
+			expectedIsValid: true,
+		},
+		{
+			// highlight-next-line
+			// Pod has labels, one has constraint that are not respected -> should be rejected
+			podLabels: map[string]string{
+				"cc-center": "team-kubewarden",
+			},
+			deniedLabels: []string{"hello"},
+			constrainedLabels: map[string]string{
+				"cc-center": `team-\d+`,
+			},
+			expectedIsValid: false,
+		},
+		{
+			// highlight-next-line
+			// Settings have a constraint, pod doesn't have this label -> should be rejected
+			podLabels: map[string]string{
+				"owner": "team-kubewarden",
+			},
+			deniedLabels: []string{"hello"},
+			constrainedLabels: map[string]string{
+				"cc-center": `team-\d+`,
+			},
+			expectedIsValid: false,
+		},
+	}
+
+	// highlight-next-line
+	// NOTE 2
+	for _, testCase := range cases {
+		basicSettings := BasicSettings{
+			DeniedLabels:      testCase.deniedLabels,
+			ConstrainedLabels: testCase.constrainedLabels,
+		}
 
-```go
-func TestEmptySettingsLeadsToRequestAccepted(t *testing.T) {
-	settings := Settings{}
+		pod := corev1.Pod{
+			Metadata: metav1.ObjectMeta{
+				Name:      "test-pod",
+				Namespace: "default",
+				Labels:    testCase.podLabels,
+			},
+		}
 
-	payload, err := kubewarden_testing.BuildValidationRequest(
-		"test_data/ingress.json",
-		&settings)
-	if err != nil {
-		t.Errorf("Unexpected error: %+v", err)
-	}
+		payload, err := kubewarden_testing.BuildValidationRequest(&pod, &basicSettings)
+		if err != nil {
+			t.Errorf("Unexpected error: %+v", err)
+		}
 
-	responsePayload, err := validate(payload)
-	if err != nil {
-		t.Errorf("Unexpected error: %+v", err)
-	}
+		responsePayload, err := validate(payload)
+		if err != nil {
+			t.Errorf("Unexpected error: %+v", err)
+		}
 
-	var response kubewarden_testing.ValidationResponse
-	if err := json.Unmarshal(responsePayload, &response); err != nil {
-		t.Errorf("Unexpected error: %+v", err)
-	}
+		var response kubewarden_protocol.ValidationResponse
+		if err := easyjson.Unmarshal(responsePayload, &response); err != nil {
+			t.Errorf("Unexpected error: %+v", err)
+		}
+
+		if testCase.expectedIsValid && !response.Accepted {
+			t.Errorf("Unexpected rejection: msg %s - code %d with test case %v",
+				*response.Message, *response.Code, testCase)
+		}
 
-	if response.Accepted != true {
-		t.Error("Unexpected rejection")
+		if !testCase.expectedIsValid && response.Accepted {
+			t.Errorf("Unexpected acceptance with test case %v", testCase)
+		}
 	}
 }
 ```
 
-As you can see we are using some test helper functions and structures provided
-by the Kubewarden SDK.
-
-The next test ensures a request can be accepted when none of its labels
-is relevant to the user:
+The test uses a "testCase" driven approach. We start by defining a `struct` that
```

Comment:
Is the word test case supposed to be capitalized like that?
```suggestion
The test uses a "testCase" driven approach. We start by defining a `struct` that
```

Reply:
No, I got tricked by the auto completion. I've fixed that.